packages/blitz-rpc/src/server-plugin.tsx

```diff
@@ -0,0 +1,88 @@
+import {getCookieParser} from "@blitzjs/auth"
+import type {BlitzServerPlugin, Middleware, Ctx, AsyncFunc, FirstParam} from "blitz"
+import {IncomingMessage, ServerResponse} from "http"
+import {DefaultOptions, QueryClient} from "react-query"
+import {DehydratedState} from "react-query/types/hydration"
+import {dehydrate, getQueryKey} from "./data-client"
+
+function ensureApiRequest(
+  req: IncomingMessage & {[key: string]: any},
+): asserts req is IncomingMessage & {cookies: {[key: string]: string}} {
+  if (!("cookies" in req)) {
+    // Cookie parser isn't include inside getServerSideProps, so we have to add it
+    req.cookies = getCookieParser(req.headers)()
+  }
+}
+function ensureMiddlewareResponse(
+  res: ServerResponse & {[key: string]: any},
+): asserts res is ServerResponse & {blitzCtx: Ctx} {
+  if (!("blitzCtx" in res)) {
+    res.blitzCtx = {} as Ctx
+  }
+}
+
+export async function setQueryClient(
+  req: IncomingMessage,
+  res: ServerResponse,
+  defaultOptions: DefaultOptions = {},
+): Promise<void> {
+  ensureApiRequest(req)
+  ensureMiddlewareResponse(res)
+
+  res.blitzCtx.prefetchBlitzQuery = async (fn, input, options = {}) => {
+    const queryClient = new QueryClient({
+      defaultOptions: {
+        mutations: {
+          ...defaultOptions.mutations,
+          ...options.mutations,
+        },
+        queries: {
+          ...defaultOptions.queries,
+          ...options.queries,
+        },
+      },
+    })
+
+    const queryKey = getQueryKey(fn, input)
+    await queryClient.prefetchQuery(queryKey, () => fn(input, res.blitzCtx))
+
+    return {
+      dehydratedState: dehydrate(queryClient),
+    }
+  }
+}
+
+export interface RpcServerPluginOptions {
+  prefetchQueryOptions?: DefaultOptions
```

Comment:
I think we can call this `reactQueryOptions` since we can use it for more than just prefetch?

Reply:
Actually, we only use it for prefetch here. User can pass react query options in **client** plugin config and we use them for all the react-query hooks. So I was thinking about the naming and figured that it can be confusing if we use the same name in both client and server.

Reply:
@beerose ok yeah that makes sense to have it named prefetch.